# HW8



Here is a Linked list class (some of which was done few labs/classes back) . We'll use a linked list as a repository for bindings. We will consider an implementation where newer bindings for existing variables occur closer to the head in the linked list. Notice that we added a `repOK` which is an identity function. Take care to see where it is used.

In [3]:
import reprlib, numbers


class LL:
    """
    Method index has been modified significantly
    It is now  parametrized with a start point and an equality function
    Additional doctests added below
    
    >>> A = LL()  
    >>> A[0]
    Traceback (most recent call last):
        ...
    IndexError: trying to index an empty LL
    >>> A.insert_front(1)
    >>> A[0]
    1
    >>> A.insert_back(2)
    >>> A[1]
    2
    >>> A
    LL([1,...])
    >>> myll = LL.from_components([1,2,2])
    >>> myll[1]
    2
    >>> len(myll)
    3
    >>> myll.index(2)
    0
    >>> myll.index(2,1)
    1
    >>> myll[3]
    Traceback (most recent call last):
        ...
    IndexError: LL index out of range
    >>> myll[0:1]
    Traceback (most recent call last):
        ...
    TypeError: LL indices must be integers
    """
    @classmethod
    def from_components(cls, components):
        inst = cls(components[0])
        for c in components[1:]:
            inst.insert_front(c)
        return inst
        
    def repOK(self, element):
        return element
    
    def __init__(self, head=None):
        if head is None:
            self._headNode = None
        else:
            head = self.repOK(head)
            self._headNode = [head, None]
            
    def insert_front(self, element):
        element = self.repOK(element)
        new_node = [element, None]
        new_node[1] = self._headNode
        self._headNode = new_node
        
    def insert_back(self, element):
        element = self.repOK(element)
        new_node = [element, None]
        curr_ptr = self._headNode
        while curr_ptr[1] is not None:
            curr_ptr = curr_ptr[1]
        curr_ptr[1]= new_node
        
    def __repr__(self):
        class_name = type(self).__name__
        if len(self)==0:
            components=""
        else:
            components = reprlib.repr(self[0])
        return '{}([{},...])'.format(class_name,components)


    def __len__(self):
        curr_ptr = self._headNode
        count = 0
        if curr_ptr==None:
            return 0
        while 1:
            count = count + 1
            if curr_ptr[1] is None:
                break
            curr_ptr = curr_ptr[1]
        return count    
    
    def __getitem__(self, index):
        class_name = type(self).__name__
        if isinstance(index, numbers.Integral): 
            curr_ptr = self._headNode
            if curr_ptr==None:
                msg = 'trying to index an empty {class_name}' 
                raise IndexError(msg.format(class_name=class_name))
            next_ptr = self._headNode[1]
            count = 0
            while 1:
                if index == count:
                    return curr_ptr[0]
                if curr_ptr[1] is None:
                    msg = '{class_name} index out of range' 
                    raise IndexError(msg.format(class_name=class_name))       
                count += 1
                curr_ptr = curr_ptr[1]
        else:
            msg = '{class_name} indices must be integers' 
            raise TypeError(msg.format(class_name=class_name))
            
            
    # index method begins search for element from start index defaulted to 0
    # allows method index to begin search from parametrized start position
    # equalityFunc with default to equality operator
    # allows subclasses to parametrize definition of equality
    def index(self, element, start=0, equalityFunc=lambda a, b : a==b):
        class_name = type(self).__name__
        curr_ptr = self._headNode
        count = 0     
        if curr_ptr==None:
            msg = 'trying to get index from empty {class_name}' 
            raise IndexError(msg.format(class_name=class_name))        
        while ((count != start) and (curr_ptr[1] != None)):           
            count += 1
            curr_ptr = curr_ptr[1]   
        if (count == start):    
            while 1:
                # comparison between curr node and element
                if equalityFunc(curr_ptr[0], element):
                    return count
                if curr_ptr[1] is None:
                    msg = '{element} is not in {class_name}' 
                    raise ValueError(msg.format(element=element, class_name=class_name))
                count += 1
                curr_ptr = curr_ptr[1]
        else:
            msg = '{element} is not in {class_name}' 
            raise ValueError(msg.format(element=element, class_name=class_name))
            
    def remove(self, element):
        class_name = type(self).__name__
        curr_ptr = self._headNode
        prev_ptr = None
        if curr_ptr==None:
            msg = 'remove from empty {class_name}' 
            raise IndexError(msg.format(class_name=class_name))
        while 1:
            if curr_ptr[0] == element:
                if prev_ptr is None:
                    self._headNode = curr_ptr[1]
                else:
                    prev_ptr[1] = curr_ptr[1]
                return None
            if curr_ptr[1] is None:
                msg = '{element} is not in {class_name}' 
                raise ValueError(msg.format(element=element, class_name=class_name))
            prev_ptr=curr_ptr
            curr_ptr = curr_ptr[1]
        
    def remove_front(self):
        class_name = type(self).__name__
        curr_ptr = self._headNode
        if curr_ptr==None:
            msg = 'remove from empty {class_name}' 
            raise IndexError(msg.format(class_name=class_name))
        self._headNode = curr_ptr[1]
        return curr_ptr[0]
    

In [4]:
from doctest import run_docstring_examples as dtest
dtest(LL, globals(), verbose=True)

Finding tests in NoName
Trying:
    A = LL()  
Expecting nothing
ok
Trying:
    A[0]
Expecting:
    Traceback (most recent call last):
        ...
    IndexError: trying to index an empty LL
ok
Trying:
    A.insert_front(1)
Expecting nothing
ok
Trying:
    A[0]
Expecting:
    1
ok
Trying:
    A.insert_back(2)
Expecting nothing
ok
Trying:
    A[1]
Expecting:
    2
ok
Trying:
    A
Expecting:
    LL([1,...])
ok
Trying:
    myll = LL.from_components([1,2,2])
Expecting nothing
ok
Trying:
    myll[1]
Expecting:
    2
ok
Trying:
    len(myll)
Expecting:
    3
ok
Trying:
    myll.index(2)
Expecting:
    0
ok
Trying:
    myll.index(2,1)
Expecting:
    1
ok
Trying:
    myll[3]
Expecting:
    Traceback (most recent call last):
        ...
    IndexError: LL index out of range
ok
Trying:
    myll[0:1]
Expecting:
    Traceback (most recent call last):
        ...
    TypeError: LL indices must be integers
ok


In [5]:
import abc
class Environment(abc.ABC):
    """
    This is the interface for an Environment. The client for 
    this interface is a language intepreter. 
    """
    @classmethod
    @abc.abstractmethod
    def empty(cls):
        return cls()
    
    @abc.abstractmethod
    def extend(self, variable, value):
        """
        extend an existing environment by binding variable to value.
        The values must be an acceptable value in the language. If the
        same variable is used twice the newer value must be bound.
        """
    
    @abc.abstractmethod
    def extend_many(self, envdict):
        """
        extend the current environment by values in the dictionary
        envdict. If the dictionary contains variables already in the
        environment, the newer values from the dictionary are bound
        """
        
    @abc.abstractmethod
    def lookup(variable):
        """
        return the unique binding of the variable and the environment it was bound
        in as a tuple. If it is not found raise a NameError as below
        """
        raise NameError("{} not found in Environment".format(variable))

### Q1

Lets inherit from class `LL` to create a TupleLL where the values in the linked list are tuples of the type `(key, value)`. 

After inheriting, add an additional method `first_match` which returns a tuple `(index_of_nearest_element, nearest element)` to the head where the first member of the tuple matches `key`. 

Also add another method `all_matches` which returns a list of tuples of allmatches, sorted by distance to head. Currently, the structure of the other methods in the linked list do not need to know about this 2-element tuple restriction on the elements. 

In [42]:
class TupleLL(LL):
    """
    A linked list whose elements must be tuples.
    
    RepInv: any insertion should leave a list with only tuples as members.
    PRE: Elements must be tuples
    
    >>> myll = TupleLL.from_components([('a', 1),('b', 2), ('hiii', 2), ('hiii', 1)])  
    >>> myll.first_match('hiii')
    (0, ('hiii', 1))
    >>> myll.first_match('b')
    (2, ('b', 2))
    >>> myll.all_matches('hiii')
    [(0, ('hiii', 1)), (1, ('hiii', 2))]

    """
    #could be implemented using a check on the whole list
    def repOK(self, element):
        assert isinstance(element, tuple), "element needs to be a tuple"
        return element
    
    # your code here
    # matching key to the first value in list element tuple
    # note the use of anonymous equality function parameter 
    # in method index inherited from class LL
    # note the use of [0] to extract the key in the tuple    
    def first_match(self, key):
        i = self.index(key, 0, lambda a, b : a[0] == b)
        matching_element = self[i]
        return (i, matching_element)
    
    # returns a list of tuples of allmatches, sorted by distance to head
    # matching key to value in the list element tuple
    # return empty list if there are no matches, in the except block
    # note i is incremented from the position of last match to look for the next match
    # append function allows list of tuples to be sorted by distance to head
    def all_matches(self, key):
        class_name = type(self).__name__
        i = 0
        matchList = []
        try: 
            while True:
                i = self.index(key, i, lambda a, b : a[0] == b )
                matching_element = self[i]
                matchList.append((i, matching_element))
                i += 1
        except ValueError:
            if (matchList == []):
                msg = '{key} is not in {class_name}' 
                raise ValueError(msg.format(key=key, class_name=class_name))
        return matchList
    
    



In [43]:
# sanity check (also included in doctest above)
myll = TupleLL.from_components([('a', 1),('b', 2), ('hiii', 2), ('hiii', 1)])
myll[0], myll[1], myll[2], myll[3]


(('hiii', 1), ('hiii', 2), ('b', 2), ('a', 1))

In [44]:
# test 1 for first_match method
myll.first_match('hiii')

(0, ('hiii', 1))

In [45]:
# test 2 for first_match method
myll.first_match('b')

(2, ('b', 2))

In [46]:
# test 3
myll.first_match('baaaaaa')

ValueError: baaaaaa is not in TupleLL

In [47]:
# sanity check (also included in doctest above)
myll.index('hiii', 1, lambda a, b : a[0] == b)

1

In [48]:
myll.all_matches('hiii')

[(0, ('hiii', 1)), (1, ('hiii', 2))]

In [49]:
myll.all_matches('a')

[(3, ('a', 1))]

In [50]:
myll.all_matches('aasdddd')

ValueError: aasdddd is not in TupleLL

In [51]:
l = TupleLL()
l.insert_front(('a', 1))
l

TupleLL([('a', 1),...])

In [52]:
l.insert_front(3)

AssertionError: element needs to be a tuple

In [53]:
from doctest import run_docstring_examples as dtest
dtest(TupleLL, globals(), verbose=True)

Finding tests in NoName
Trying:
    myll = TupleLL.from_components([('a', 1),('b', 2), ('hiii', 2), ('hiii', 1)])  
Expecting nothing
ok
Trying:
    myll.first_match('hiii')
Expecting:
    (0, ('hiii', 1))
ok
Trying:
    myll.first_match('b')
Expecting:
    (2, ('b', 2))
ok
Trying:
    myll.all_matches('hiii')
Expecting:
    [(0, ('hiii', 1)), (1, ('hiii', 2))]
ok


### Q2.

Implement a inplementation class `Env2` that impements the environment interface using the `TupleLL`. Let us ask what the AbsFun and RepInv are for this implementation. Write an AbsFun for it, noting that in this concrete representation, there can be multiple key-value pairs in the environment for the same key, and that there is a way to disambiguate the correct one. 

(In general, the Absfun should limit the keys and values allowed for our bindings (as according to our language) as well. But we wont do that here.)




In [105]:
#import LL, TupleLL

import reprlib
class Env2(Environment):
    """
    Implements ABC Environment
    
    AbsFun: The list [(key1, value1), (key2, value2), ...] represents the key-value pairs 
    in the environment, this is implemented in the __repr__ function
    
    The list may contain multiple key-value pairs in the environment for the same key.
    
    The way to disambiguate the correct one will be the one closer to the head, i.e. 
    closer to the start of the list
    
    >>> e = Env2()
    >>> dict = {'Maria': 1, 'James': 's', 'Donald': 10, 'Melissa': [1, 1, 2]}
    >>> e.extend_many(dict) 
    >>> e
    [('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]
    >>> e.extend('Donald', 100)
    >>> e
    [('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]
    >>> e.lookup('Donald')
    (100, TupleLL([('Donald', 100),...]))
    
    """
    @classmethod
    def empty(cls):
        return cls()
    
    # underlying data structure is TupleLL
    def __init__(self):
        self._bindingList = TupleLL()
        self.class_name = type(self).__name__
    
    # if the binding list is < 10, print all the tuples
    # otherwise, print the first 10 elements
    def __repr__(self):
        outList = []
        for i in range(min(len(self._bindingList), 10)):
            outList.append(self._bindingList[i])
        s = "{}".format(outList)
        if len(self._bindingList)>10:
            s = s + " ..."
        return s
    
    def __str__(self):
        outList = []
        s = "The environment is "
        for i in range(min(len(self._bindingList), 10)):
            outList.append(self._bindingList[i])
        s = s + "{}".format(outList)
        if len(self._bindingList)>10:
            s = s + " ..."
        return s
    
    def extend(self, variable, value):
        """
        extend an existing environment by binding variable to value.
        The values must be an acceptable value in the language. If the
        same variable is used twice the newer value must be bound.
        """
        self._bindingList.insert_front((variable, value))
    
    def extend_many(self, envdict):
        """
        extend the current environment by values in the dictionary
        envdict. If the dictionary contains variables already in the
        environment, the newer values from the dictionary are bound
        """
        for key in envdict:
            t = key, envdict[key]
            self._bindingList.insert_front(t)
        #self._bindingList.insert_front((key, envdict[key]) for key in envdict)
        
    def lookup(self, variable):
        """
        return the unique binding of the variable and the environment it was bound
        in as a tuple. If it is not found raise a NameError as below
        """
        try:
            return (self._bindingList.first_match(variable)[1][1], self._bindingList)
        except ValueError:
            raise NameError("{} not found in Environment".format(variable))
    

In [106]:
# sanity check (also included in doctest)
e = Env2()
dict = {'Maria': 1, 'James': 's', 'Donald': 10, 'Melissa': [1, 1, 2]}


In [107]:
e.extend_many(dict) 
e

[('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]

In [108]:
e.extend('Donald', 100)
e

[('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]

In [109]:
print (e)

The environment is [('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]


In [110]:
e.extend('Maria', 100)

In [111]:
e

[('Maria', 100), ('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]

In [112]:
e.lookup('Donald')

(100, TupleLL([('Maria', 100),...]))

In [113]:
from doctest import run_docstring_examples as dtest
dtest(Env2, globals(), verbose=True)

Finding tests in NoName
Trying:
    e = Env2()
Expecting nothing
ok
Trying:
    dict = {'Maria': 1, 'James': 's', 'Donald': 10, 'Melissa': [1, 1, 2]}
Expecting nothing
ok
Trying:
    e.extend_many(dict) 
Expecting nothing
ok
Trying:
    e
Expecting:
    [('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]
ok
Trying:
    e.extend('Donald', 100)
Expecting nothing
ok
Trying:
    e
Expecting:
    [('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]
ok
Trying:
    e.lookup('Donald')
Expecting:
    (100, TupleLL([('Donald', 100),...]))
ok


### Q3. `repinv`

Now let us write and test a RepInv for this implementation. Note that just like in the case with the list-with-duplicates implementation of a set, its hard to come up with any uniqueness-of-binding RepInv. But we can ask the following: a newer binding overrides an old one. Write a Repinv which implements this and include a repOK methodwhich makes sure that any methods mutating the environment respect this RepInv and any observers return the correct binding. 

Hint: the `repOK` will need a signature `def repOK(self, key, value)` and also be used in the `lookup` method. It will use all_matches, and not be returning anything.

In [126]:
#your code here
# copy of code from Q2 with method repOK and its usage within the class added

#import LL, TupleLL

import reprlib
class Env2(Environment):
    """
    Implements ABC Environment
    
    AbsFun: The list [(key1, value1), (key2, value2), ...] represents the key-value pairs 
    in the environment, this is implemented in the __repr__ function
    
    The list may contain multiple key-value pairs in the environment for the same key.
    
    The way to disambiguate the correct one will be the one closer to the head, i.e. 
    closer to the start of the list
    
    >>> e = Env2()
    >>> dict = {'Maria': 1, 'James': 's', 'Donald': 10, 'Melissa': [1, 1, 2]}
    >>> e.extend_many(dict) 
    >>> e.extend('Donald', 100)
    >>> e
    [('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]
    >>> e.lookup('Donald')
    (100, TupleLL([('Donald', 100),...]))
    
    """
    @classmethod
    def empty(cls):
        return cls()
    
    # here is an implementation of repinv
    def repOK(self, key, value):
        assert (key != None), "Key cannot be none"
        assert (value != None), "Value cannot be none"
        # find if the key has duplicate values in the environment
        allMatchList = self._bindingList.all_matches(key)
        if (len(allMatchList) > 1):
            # return the value at minimum index for duplicate keys
            val_fromMatchList = min(allMatchList)[1][1]
            # return the lookup value for the same key
            val_fromlookup = self.lookup(key)[0]
            # if the two values are not equal
            # repinv is invalid
            assert (val_fromMatchList == val_fromlookup), "Representation Integrity Not Valid, lookup is not returning latest value of key"
                
                
    # underlying data structure is TupleLL
    def __init__(self):
        self._bindingList = TupleLL()
        self.class_name = type(self).__name__
    
    # if the binding list is < 10, print all the tuples
    # otherwise, print the first 10 elements
    def __repr__(self):
        outList = []
        for i in range(min(len(self._bindingList), 10)):
            outList.append(self._bindingList[i])
        s = "{}".format(outList)
        if len(self._bindingList)>10:
            s = s + " ..."
        return s
    
    def __str__(self):
        outList = []
        s = "The environment is "
        for i in range(min(len(self._bindingList), 10)):
            outList.append(self._bindingList[i])
        s = s + "{}".format(outList)
        if len(self._bindingList)>10:
            s = s + " ..."
        return s
    
    def extend(self, variable, value):
        """
        extend an existing environment by binding variable to value.
        The values must be an acceptable value in the language. If the
        same variable is used twice the newer value must be bound.
        """
        self._bindingList.insert_front((variable, value))
        self.repOK(variable, value)
    
    def extend_many(self, envdict):
        """
        extend the current environment by values in the dictionary
        envdict. If the dictionary contains variables already in the
        environment, the newer values from the dictionary are bound
        """
        for key in envdict:
            t = key, envdict[key]
            self._bindingList.insert_front(t)
            self.repOK(key, envdict[key])
        
    def lookup(self, variable):
        """
        return the unique binding of the variable (value) and the environment it was bound
        in as a tuple. If it is not found raise a NameError as below
        """
        try:
            tup = self._bindingList.first_match(variable)
            return (tup[1][1], self._bindingList)
        except ValueError:
            raise NameError("{} not found in Environment".format(variable))


In [127]:
testEnv = Env2()
dict = {'Maria': 1, 'James': 's', 'Donald': 10, 'Melissa': [1, 1, 2]}


In [128]:
testEnv.extend_many(dict)

In [129]:
testEnv

[('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]

In [130]:
testEnv.extend('Donald', 100)

In [131]:
from doctest import run_docstring_examples as dtest
dtest(Env2, globals(), verbose=True)

Finding tests in NoName
Trying:
    e = Env2()
Expecting nothing
ok
Trying:
    dict = {'Maria': 1, 'James': 's', 'Donald': 10, 'Melissa': [1, 1, 2]}
Expecting nothing
ok
Trying:
    e.extend_many(dict) 
Expecting nothing
ok
Trying:
    e.extend('Donald', 100)
Expecting nothing
ok
Trying:
    e
Expecting:
    [('Donald', 100), ('Donald', 10), ('Melissa', [1, 1, 2]), ('Maria', 1), ('James', 's')]
ok
Trying:
    e.lookup('Donald')
Expecting:
    (100, TupleLL([('Donald', 100),...]))
ok


### Q4. Running the program

Just make sure that the program actually runs :-). Note some changes here: we are using `yield`s instead of `return`s at multiple places.

Lets use this class in a function which creates a global environment for our calculator referenced as `globenv`. we first register this implementation for the `Environment` interface

In [132]:
Environment.register(Env2)

__main__.Env2

In [133]:
import math
def global_env(envclass):
    "An environment with some Scheme standard procedures."
    import math, operator as op
    env = envclass.empty()
    env.extend_many(vars(math))
    env.extend_many({
        '+':op.add, '-':op.sub, '*':op.mul, '/':op.truediv, 
        'abs':     abs,
        'max':     max,
        'min':     min,
        'round':   round,
        '>':op.gt, '<':op.lt, '>=':op.ge, '<=':op.le, '==':op.eq,
        'not':     op.not_
    })
    return env

In [134]:
class Function():
    def __init__(self, params, parsedbody, env):
        self.params = params
        self.code = parsedbody
        self.env = env
        self.envclass = env.__class__
        
    def __call__(self, *args):
        funcenv = self.envclass(outerenv = self.env)
        funcenv.extend_many(zip(self.params, args))
        return eval_ptree(self.code, funcenv)

In [135]:
globenv = global_env(Env2)

In [136]:
def typer(token):
    try:
        t = int(token)
        return t
    except ValueError:
        try:
            t = float(token)
            return t
        except ValueError:
            return str(token)
        
def lex(loc):
    tokenlist =  loc.replace('(', ' ( ').replace(')', ' ) ').split()
    return [typer(t) for t in tokenlist]

def syn(tokens):
    if len(tokens) == 0:
        return []
    token = tokens.pop(0)
    if token == '(':
        L = []
        while tokens[0] != ')':
            L.append(syn(tokens))
        tokens.pop(0) # pop off ')'
        return L
    else:
        if token==')':
            assert 1, "should not have got here"
        return token
    
def parse(loc):
    return syn(lex(loc))

In [137]:
def eval_ptree(x, env):
    Symbol = str
    fmap={'#t':True, '#f':False, 'nil':None}
    if x in ('#t', '#f', 'nil'):
        return fmap[x]        
    elif isinstance(x, Symbol):
        # variable lookup
        return env.lookup(x)[0]
    elif not isinstance(x, list):  # constant
        return x
    elif len(x)==0: #noop
        return None
    elif x[0]=='if':
        (_, predicate, truexpr, falseexpr) = x
        if eval_ptree(predicate, env):
            expression = truexpr
        else:
            expression = falseexpr
        return eval_ptree(expression, env)
    elif x[0] == 'def':         # variable definition
        (_, var, expression) = x
        #postorder traversal by nested eval is needed below
        # your code here
        env.extend(var, eval_ptree(expression, env))
    elif x[0] == 'store':           # (set! var exp)
        (_, var, exp) = x
        env.lookup(var)[1].extend(var, eval_ptree(exp, env))
    elif x[0] == 'func':
        (_, parameters, parsedbody) = x
        return Function(parameters, parsedbody, env)
    else:                          # operator
        op = eval_ptree(x[0], env)
        #postorder traversal to get subexpressione before running the op
        args = [eval_ptree(arg, env) for arg in x[1:]]
        return op(*args)

In [138]:
class Program():
    
    def __init__(self, program, env):
        self.program = [e.strip() for e in program.split('\n')]
        self.env = env
        
    def __iter__(self):
        for line in self.program:
            yield line
    
    def parse(self):
        for l in iter(self):
            yield parse(l)
            
    def run(self):
        for l in iter(self):
            yield eval_ptree(parse(l), self.env)

In [139]:
program = """
(def radius 5)
(* pi (* radius radius))
"""

In [140]:
p=Program(program, globenv)
list(iter(p))

['', '(def radius 5)', '(* pi (* radius radius))', '']

In [141]:
for s in p.parse():
    print(s)

[]
['def', 'radius', 5]
['*', 'pi', ['*', 'radius', 'radius']]
[]


In [143]:
for result in p.run():
    print(result)

None
None
78.53981633974483
None


We see that 78.5 is indeed the result of pi multiplied by square of r.